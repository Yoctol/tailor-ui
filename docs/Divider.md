---
id: divider
title: Divider
---

## Examples

```js
import { Divider } from 'tailor-ui';
```

### Basic

```jsx live
<>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider>With Text</Divider>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider dashed />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
</>
```

### Vertical

```jsx live
<>
  Text
  <Divider type="vertical" />
  <a href="#">Link</a>
  <Divider type="vertical" />
  <a href="#">Link</a>
</>
```

### Orientation of title

```jsx live
<>
  <Divider orientation="left">Left Text</Divider>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
    merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo
    modo.
  </p>
  <Divider orientation="right">Right Text</Divider>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
    merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo
    modo.
  </p>
</>
```

## API
