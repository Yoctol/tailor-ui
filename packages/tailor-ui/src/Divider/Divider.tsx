import React, { FunctionComponent } from 'react';
import styled, { css } from 'styled-components';
import {
  ColorProps,
  FontSizeProps,
  FontWeightProps,
  SpaceProps,
  color,
  fontSize,
  fontWeight,
  space,
} from 'styled-system';

type IStyledDividerProps = SpaceProps &
  ColorProps &
  FontSizeProps &
  FontWeightProps & {
    type: 'horizontal' | 'vertical';
    orientation?: 'left' | 'right';
    dashed?: boolean;
    withText: boolean;
  };

const StyledInnerText = styled.span`
  display: inline-block;
  padding: 0 ${p => p.theme.space[4]};
`;

const StyledDivider = styled.div<IStyledDividerProps>`
  background-color: ${p => p.theme.colors.gray300};

  ${p => {
    if (p.type === 'vertical') {
      return css`
        display: inline-block;
        position: relative;
        top: -0.06em;
        width: 1px;
        height: 0.9em;
        vertical-align: middle;

        ${p.withText &&
          css`
            border-top: 0;
            &::before,
            &::after {
              border-style: dashed none none;
            }
          `}
      `;
    }

    return css`
      display: block;
      width: 100%;
      min-width: 100%;
      height: 1px;
      margin: ${p.theme.space[4]} 0;
      clear: both;

      ${p.withText &&
        css`
          display: table;
          background-color: transparent;
          text-align: center;
          white-space: nowrap;

          &::before,
          &::after {
            content: '';
            display: table-cell;
            position: relative;
            top: 50%;
            width: 50%;
            border-top: ${p.theme.borders.base};
            border-color: ${p.theme.colors.gray300};
            transform: translateY(50%);
          }

          ${p.orientation &&
            css`
              &::before {
                top: 50%;
                width: ${p.orientation === 'left' ? '5%' : '95%'};
              }

              &::after {
                top: 50%;
                width: ${p.orientation === 'left' ? '95%' : '5%'};
              }

              ${StyledInnerText} {
                display: inline-block;
                padding: 0 10px;
              }
            `}
        `}
    `;
  }}

  ${p =>
    p.dashed &&
    css`
      border-top: ${p.theme.borders.dashed};
      border-color: ${p.theme.colors.gray300};
      background-color: transparent;
    `}

    ${space};
    ${color};
    ${fontSize};
    ${fontWeight};
`;

StyledDivider.defaultProps = {
  color: 'gray400',
  fontSize: 'sm',
  fontWeight: 500,
};

export interface DividerProps {
  /**
   * direction type of divider
   */
  type?: 'horizontal' | 'vertical';
  /**
   * position of title inside divider
   */
  orientation?: 'left' | 'right';
  /**
   * 	whether line is dashed
   */
  dashed?: boolean;
}

const Divider: FunctionComponent<DividerProps> = ({
  type = 'horizontal',
  orientation,
  dashed,
  children,
  ...props
}) => (
  <StyledDivider
    type={type}
    orientation={orientation}
    dashed={dashed}
    withText={!!children}
    my={type === 'horizontal' ? 2 : 0}
    mx={type === 'vertical' ? 2 : 0}
    {...props}
  >
    {children && <StyledInnerText>{children}</StyledInnerText>}
  </StyledDivider>
);

export default Divider;
