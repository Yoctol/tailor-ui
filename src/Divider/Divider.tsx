import React, { FunctionComponent } from 'react';
import styled, { css } from 'styled-components';

import tag from 'utils/CleanTag';

interface IStyledDividerProps {
  type: 'horizontal' | 'vertical';
  orientation?: 'left' | 'right';
  dashed?: boolean;
  withText: boolean;
}

const StyledInnerText = styled(tag.span)`
  display: inline-block;
  padding: 0 ${p => p.theme.space[3]};
`;

const StyledDivider = styled<IStyledDividerProps>(tag.div)`
  background-color: ${p => p.theme.colors.gray300};

  ${p => {
    if (p.type === 'vertical') {
      return css`
        display: inline-block;
        position: relative;
        top: -0.06em;
        width: 1px;
        height: 0.9em;
        margin: 0 8px;
        vertical-align: middle;

        ${p.withText &&
          css`
            border-top: 0;
            &::before,
            &::after {
              border-style: dashed none none;
            }
          `}
      `;
    }

    return css`
      display: block;
      width: 100%;
      min-width: 100%;
      height: 1px;
      margin: ${p.theme.space[3]} 0;
      clear: both;

      ${p.withText &&
        css`
          display: table;
          margin: 16px 0;
          background-color: transparent;
          color: ${p.theme.colors.primaryDark};
          font-size: ${p.theme.fontSizes.lg};
          font-weight: 500;
          text-align: center;
          white-space: nowrap;

          &::before,
          &::after {
            content: '';
            display: table-cell;
            position: relative;
            top: 50%;
            width: 50%;
            border-top: ${p.theme.borders.base};
            border-color: ${p.theme.colors.gray300};
            transform: translateY(50%);
          }

          ${p.orientation &&
            css`
              &::before {
                top: 50%;
                width: ${p.orientation === 'left' ? '5%' : '95%'};
              }

              &::after {
                top: 50%;
                width: ${p.orientation === 'left' ? '95%' : '5%'};
              }

              ${StyledInnerText} {
                display: inline-block;
                padding: 0 10px;
              }
            `}
        `}
    `;
  }}

  ${p =>
    p.dashed &&
    css`
      border-top: ${p.theme.borders.dashed};
      border-color: ${p.theme.colors.gray300};
      background-color: transparent;
    `}
`;

export interface IDividerProps {
  /**
   * direction type of divider
   */
  type?: 'horizontal' | 'vertical';
  /**
   * position of title inside divider
   */
  orientation?: 'left' | 'right';
  /**
   * 	whether line is dashed
   */
  dashed?: boolean;
}

const Divider: FunctionComponent<IDividerProps> = ({
  type = 'horizontal',
  orientation,
  dashed,
  children,
}) => (
  <StyledDivider
    type={type}
    orientation={orientation}
    dashed={dashed}
    withText={!!children}
  >
    {children && <StyledInnerText>{children}</StyledInnerText>}
  </StyledDivider>
);

export default Divider;
