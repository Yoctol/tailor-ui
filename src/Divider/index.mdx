---
name: Divider
menu: Components
route: /components/divider
---

import { PropsTable, Playground } from 'docz';
import Divider from './';

# Divider

## Examples

```js
import { Divider } from '@yoctol/ui';
```

### Basic

<Playground>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider>With Text</Divider>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <Divider dashed />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti
  licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
</Playground>

### Vertical

<Playground>
  Text
  <Divider type="vertical" />
  <a href="#">Link</a>
  <Divider type="vertical" />
  <a href="#">Link</a>
</Playground>

### Orientation of title

<Playground>
  <Divider orientation="left">Left Text</Divider>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
    merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo
    modo.
  </p>
  <Divider orientation="right">Right Text</Divider>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
    merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo
    modo.
  </p>
</Playground>

## API

<PropsTable of={Divider} />
