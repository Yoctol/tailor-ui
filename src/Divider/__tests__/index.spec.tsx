// eslint
import React from 'react';

import { render } from 'test/test-utils';

import Divider from '..';

describe('Divider', () => {
  it('should render correctly', () => {
    const { container } = render(
      <div>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
        merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen,
        quo modo.
        <Divider />
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
        merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen,
        quo modo.
        <Divider>With Text</Divider>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
        merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen,
        quo modo.
        <Divider dashed />
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
        merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen,
        quo modo.
      </div>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correct with type vertical', () => {
    const { container } = render(
      <div>
        Text
        <Divider type="vertical" />
        {/* eslint-disable-next-line */}
        <a href="#">Link</a>
        <Divider type="vertical" />
        {/* eslint-disable-next-line */}
        <a href="#">Link</a>
      </div>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correct with orientation', () => {
    const { container } = render(
      <div>
        <Divider orientation="left">Left Text</Divider>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
          merninisti licere mihi ista probare, quae sunt a te dicta? Refert
          tamen, quo modo.
        </p>
        <Divider orientation="right">Right Text</Divider>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne
          merninisti licere mihi ista probare, quae sunt a te dicta? Refert
          tamen, quo modo.
        </p>
      </div>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
