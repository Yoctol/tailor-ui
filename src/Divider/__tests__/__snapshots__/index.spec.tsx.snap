// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Divider should render correct with orientation 1`] = `
.c2 {
  display: inline-block;
  padding: 0 1rem;
}

.c0 {
  background-color: #e0e0e0;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 24px 0;
  clear: both;
  display: table;
  margin: 16px 0;
  background-color: transparent;
  color: #191539;
  font-size: 1.125rem;
  font-weight: 500;
  text-align: center;
  white-space: nowrap;
}

.c0::before,
.c0::after {
  content: '';
  display: table-cell;
  position: relative;
  top: 50%;
  width: 50%;
  border-top: 1px solid;
  border-color: #e0e0e0;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c0::before {
  top: 50%;
  width: 5%;
}

.c0::after {
  top: 50%;
  width: 95%;
}

.c0 .c1 {
  display: inline-block;
  padding: 0 10px;
}

.c3 {
  background-color: #e0e0e0;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 24px 0;
  clear: both;
  display: table;
  margin: 16px 0;
  background-color: transparent;
  color: #191539;
  font-size: 1.125rem;
  font-weight: 500;
  text-align: center;
  white-space: nowrap;
}

.c3::before,
.c3::after {
  content: '';
  display: table-cell;
  position: relative;
  top: 50%;
  width: 50%;
  border-top: 1px solid;
  border-color: #e0e0e0;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c3::before {
  top: 50%;
  width: 95%;
}

.c3::after {
  top: 50%;
  width: 5%;
}

.c3 .c1 {
  display: inline-block;
  padding: 0 10px;
}

<div>
  <div
    class="c0"
    orientation="left"
    type="horizontal"
  >
    <span
      class="c1 c2"
    >
      Left Text
    </span>
  </div>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  </p>
  <div
    class="c3"
    orientation="right"
    type="horizontal"
  >
    <span
      class="c1 c2"
    >
      Right Text
    </span>
  </div>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  </p>
</div>
`;

exports[`Divider should render correct with type vertical 1`] = `
.c0 {
  background-color: #e0e0e0;
  display: inline-block;
  position: relative;
  top: -0.06em;
  width: 1px;
  height: 0.9em;
  margin: 0 8px;
  vertical-align: middle;
}

<div>
  Text
  <div
    class="c0"
    type="vertical"
  />
  <a
    href="#"
  >
    Link
  </a>
  <div
    class="c0"
    type="vertical"
  />
  <a
    href="#"
  >
    Link
  </a>
</div>
`;

exports[`Divider should render correctly 1`] = `
.c2 {
  display: inline-block;
  padding: 0 1rem;
}

.c0 {
  background-color: #e0e0e0;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 24px 0;
  clear: both;
}

.c1 {
  background-color: #e0e0e0;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 24px 0;
  clear: both;
  display: table;
  margin: 16px 0;
  background-color: transparent;
  color: #191539;
  font-size: 1.125rem;
  font-weight: 500;
  text-align: center;
  white-space: nowrap;
}

.c1::before,
.c1::after {
  content: '';
  display: table-cell;
  position: relative;
  top: 50%;
  width: 50%;
  border-top: 1px solid;
  border-color: #e0e0e0;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c3 {
  background-color: #e0e0e0;
  display: block;
  width: 100%;
  min-width: 100%;
  height: 1px;
  margin: 24px 0;
  clear: both;
  border-top: 1px dashed;
  border-color: #e0e0e0;
  background-color: transparent;
}

<div>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <div
    class="c0"
    type="horizontal"
  />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <div
    class="c1"
    type="horizontal"
  >
    <span
      class="c2"
    >
      With Text
    </span>
  </div>
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
  <div
    class="c3"
    type="horizontal"
  />
  Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nonne merninisti licere mihi ista probare, quae sunt a te dicta? Refert tamen, quo modo.
</div>
`;
